metaflow/extension_support.py

```diff
@@ -0,0 +1,441 @@
+                                    )
+                            else:
+                                _ext_debug(
+                                    "\tExtends '%s' with config '%s'"
+                                    % (_extension_points[idx], config_module)
+                                )
+                                extension_points_to_pkg[_extension_points[idx]][
+                                    dist.metadata["Name"]
+                                ] = MFExtPackage(
+                                    package_name=dist.metadata["Name"],
+                                    tl_package=parts[1],
+                                    config_module=config_module,
+                                )
+                            break
+    extension_points_to_pkg.default_factory = list
+
+    # At this point, we have all the packages that contribute to EXT_PKG,
+    # we now check to see if there is an order to respect based on dependencies. We will
+    # return an ordered list that respects that order and is ordered alphabetically in
+    # case of ties.
+    pkg_to_reqs_count = {}
+    req_to_dep = {}
+    mf_ext_packages_set = set(mf_ext_packages)
+    for pkg_name in mf_ext_packages:
+        req_count = 0
+        req_pkgs = [x.split()[0] for x in metadata.requires(pkg_name)]
+        for req_pkg in req_pkgs:
+            if req_pkg in mf_ext_packages_set:
+                req_count += 1
+                req_to_dep.setdefault(req_pkg, []).append(pkg_name)
+        pkg_to_reqs_count[pkg_name] = req_count
+
+    # Find roots
+    mf_pkg_list = []
+    to_process = []
+    for pkg_name, count in pkg_to_reqs_count.items():
+        if count == 0:
+            to_process.append(pkg_name)
+
+    # Add them in alphabetical order
+    to_process.sort()
+    mf_pkg_list.extend(to_process)
+    # Find rest topologically
+    while to_process:
+        next_round = []
+        for pkg_name in to_process:
+            del pkg_to_reqs_count[pkg_name]
+            for dep in req_to_dep.get(pkg_name, []):
+                cur_req_count = pkg_to_reqs_count[dep]
+                if cur_req_count == 1:
+                    next_round.append(dep)
+                else:
+                    pkg_to_reqs_count[dep] = cur_req_count - 1
+        # Add those in alphabetical order
+        next_round.sort()
+        mf_pkg_list.extend(next_round)
+        to_process = next_round
+
+    # Check that we got them all
+    if len(pkg_to_reqs_count) > 0:
+        raise RuntimeError(
+            "Unresolved dependencies in %s: %s" % (EXT_PKG, str(pkg_to_reqs_count))
+        )
+
+    # Figure out the per extension point order
+    for k, v in extension_points_to_pkg.items():
+        l = [v[pkg] for pkg in mf_pkg_list if pkg in v]
+        extension_points_to_pkg[k] = l
+    return mf_pkg_list, extension_points_to_pkg
+
+
+_all_packages, _pkgs_per_extension_point = _get_extension_packages()
```

Comment:
`_pkgs_per_extension_point` is currently a dict of the form : 
```python
{
    "plugins.cards":[
        MFExtPackage(package_name='metaflow-ext-testing', tl_package='somedevorg', config_module=None), 
    ]
}
```

Can we change this to : 

```python
{
    "plugins.cards":{
        "somedevorg":[ # "somedevorg" is a top level module
            MFExtPackage(package_name='metaflow-ext-testing', tl_package='somedevorg', config_module=None), 
            MFExtPackage(package_name='metaflow-ext-testing', tl_package='somedevorg2', config_module=None)
        ]
    }
}
```

### Why make this change? 

Conside the following directory structure : 


```
your_package/ # the name of this dir doesn't matter
├ setup.py
├ metaflow_ext/ # namespace package name (This will change to `metaflow_extensions`
│  └ organizationA/      # package to keep org level extensions seperate ; # dir name must match the package name in `setup.py`
│      ├__init__.py    # This will have an __init__.py file
│      └ plugins/ # NO __init__.py file
│        └ cards/ # NO __init__.py file # This is a namespace package. 
│           └ my_card_module/  # Name of card_module
│               └ __init__.py. # This is the __init__.py is required to recoginize `my_card_module` as a package
│               └ somerandomfile.py. # Some file as a part of the package. 
│  └ organizationB/      
│      ├__init__.py    
│      └ plugins/ 
│        └ cards/ 
│           └ my_card_module_2/  # Name of card_module
│               └ __init__.py. # This is the __init__.py is required to recoginize `my_card_module_2` as a package
│               └ somerandomfile_2.py. # Some file as a part of the package. 
.
```

When I install the `setup.py` here, the current method will only recognize the top level module `organizationA` and  not `organizationB`

Reply:
Fair point. I had not considered this setup. I will integrate your suggestion.